lib/auth/github.go

```diff
@@ -84,6 +91,95 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
 	return nil
 }
 
+// HTTPRequester allows a net/http.Client to be mocked for tests.
+type HTTPRequester interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// checkGithubFeatureSupport returns an error if any of the Github
+// organizations specified in this connector use external SSO. SSO
+// is a Teleport Enterprise feature that should not be allowed for OSS.
+// If client is nil a net/http.Client will be used.
+func checkGithubFeatureSupport(conn types.GithubConnector, client HTTPRequester) error {
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	if client == nil {
+		transport, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		client = &http.Client{
+			Transport: transport,
+		}
+	}
+
+	checkOrg := func(org string) error {
+		// A Github organization will have a "sso" page reachable if it
+		// supports external SSO. There doesn't seem to be any way to get this
+		// information from the Github REST API without being an owner of the
+		// Github organization, so check if this exists instead.
+		ssoURL := fmt.Sprintf("%s/%s/%s", githubOrgsURL, url.PathEscape(org), "sso")
+
+		const retries int = 3
+		var resp *http.Response
+		for i := 0; i < retries; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), defaults.HTTPRequestTimeout)
+			defer cancel()
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			var urlErr *url.Error
+			resp, err = client.Do(req)
+			if err == nil {
+				break
+			} else if errors.As(err, &urlErr) && urlErr.Timeout() {
+				if i == retries-1 {
+					// The connection timed out a couple of times in a row,
+					// stop trying and return the error.
+					return trace.ConnectionProblem(err, "Timed out trying to reach GitHub to check for organization external SSO.")
+				}
+				// Connection timed out, try to make the request again
+				continue
+			}
+			// Unknown error, don't try making any more requests
+			return trace.Wrap(err, "Unknown error trying to reach GitHub to check for organization external SSO")
+		}
+		resp.Body.Close()
+
+		if resp.StatusCode == 200 {
+			return trace.AccessDenied(
+				"GitHub organization %s uses external SSO, please purchase a Teleport Enterprise license if you want to authenticate with this organization",
+				org,
+			)
+		}
+
+		return nil
+	}
+
+	// Check each organization only once
+	orgs := make(map[string]struct{})
+	for _, mapping := range conn.GetTeamsToLogins() {
+		orgs[mapping.Organization] = struct{}{}
+	}
+	for _, mapping := range conn.GetTeamsToRoles() {
+		orgs[mapping.Organization] = struct{}{}
+	}
+
+	for org := range orgs {
+		if err := checkOrg(org); err != nil {
+			return trace.Wrap(err)
+		}
+	}
```

Comment:
So looks like we're checking every org in the connector on every login? Shouldn't we instead check the orgs that the user who's logging in belongs to?

Reply:
Done